rollup/cmd/rollup_relayer/app/app.go

```diff
@@ -132,3 +148,319 @@ func Run() {
 		os.Exit(1)
 	}
 }
+
+func restoreFullPreviousState(cfg *config.Config, db *gorm.DB, chunkProposer *watcher.ChunkProposer, batchProposer *watcher.BatchProposer, bundleProposer *watcher.BundleProposer, l2Watcher *watcher.L2WatcherClient) error {
+	log.Info("Restoring full previous state with", "L1 block height", cfg.RecoveryConfig.L1BlockHeight, "latest finalized batch", cfg.RecoveryConfig.LatestFinalizedBatch)
+
+	// DB state should be clean: the latest batch in the DB should be finalized on L1. This function will
+	// restore all batches between the latest finalized batch in the DB and the latest finalized batch on L1.
+
+	// 1. Get latest finalized batch stored in DB
+	latestDBBatch, err := batchProposer.BatchORM().GetLatestBatch(context.Background())
+	if err != nil {
+		return fmt.Errorf("failed to get latest batch from DB: %w", err)
+	}
+	fmt.Println("latestDBBatch", latestDBBatch)
+
+	// TODO:
+	//  1. what if it is a fresh start? -> latest batch is nil
+	//latestDBBatch.CommitTxHash
+
+	log.Info("Latest finalized batch in DB", "batch", latestDBBatch.Index, "hash", latestDBBatch.Hash)
+
+	// TODO: make these parameters -> part of genesis config?
+	scrollChainAddress := common.HexToAddress("0x2D567EcE699Eabe5afCd141eDB7A4f2D0D6ce8a0")
+	l1MessageQueueAddress := common.HexToAddress("0xF0B2293F5D834eAe920c6974D50957A1732de763")
+
+	l1Client, err := ethclient.Dial(cfg.L1Config.Endpoint)
+	if err != nil {
+		return fmt.Errorf("failed to connect to L1 client: %w", err)
+	}
+	reader, err := l1.NewReader(context.Background(), l1.Config{
+		ScrollChainAddress:    scrollChainAddress,
+		L1MessageQueueAddress: l1MessageQueueAddress,
+	}, l1Client)
+	if err != nil {
+		return fmt.Errorf("failed to create L1 reader: %w", err)
+	}
+
+	// 2. Get latest finalized L1 block
+	latestFinalizedL1Block, err := reader.GetLatestFinalizedBlockNumber()
+	if err != nil {
+		return fmt.Errorf("failed to get latest finalized L1 block number: %w", err)
+	}
+
+	log.Info("Latest finalized L1 block number", "latest finalized L1 block", latestFinalizedL1Block)
+
+	// 3. Get latest finalized batch from contract (at latest finalized L1 block)
+	latestFinalizedBatch, err := reader.LatestFinalizedBatch(latestFinalizedL1Block)
+	if err != nil {
+		return fmt.Errorf("failed to get latest finalized batch: %w", err)
+	}
+
+	log.Info("Latest finalized batch from L1 contract", "latest finalized batch", latestFinalizedBatch, "at latest finalized L1 block", latestFinalizedL1Block)
+
+	// 4. Get batches one by one from stored in DB to latest finalized batch.
+	receipt, err := l1Client.TransactionReceipt(context.Background(), common.HexToHash(latestDBBatch.CommitTxHash))
```

Comment:
due to the pruning of txn index, not sure if we may need an archive node or fallback to fetch the whole block for this.